### DISCUSSION TITLE
What to do with mutants in custom GetHashCode() implementations?

### DISCUSSION THREAD

**Discussion by psfinaki** [stryker-mutator/stryker-net#1667] [Q&A] [answered]
We are trying to get high mutation scores in our code so there'll be more questions from me :D 

So, it's a common practice to implement `GetHashCode` in classes and it often looks like this (code from here):
```csharp
public override int GetHashCode()
{
   unchecked
   {
      return (Property1.GetHashCode() * 397) ^ Property2.GetHashCode();
   }
}
```
Such pattern is recommended even by holy Jon Skeet [here](https://stackoverflow.com/a/1646913/3232646) and [here's](https://stackoverflow.com/a/102878/3232646) the reason for 397 if you wonder :D

Now apparently Stryker mutates in like this:
```csharp
public override int GetHashCode()
{
   unchecked
   {
      return ~((Property1.GetHashCode() * 397) ^ Property2.GetHashCode());
   }
}
```

Do you have any particular recommendation on how to handle this?

@xakep139

**Comment by dupdob** [comment, marked as answer]
I am close to 100% with [NFluent](https://github.com/dupdob/NFluent). I do have unit tests for the couple for `GetHashCode` overrides, if that is your question.
Those have to be updated sometimes as hashcode calculation may change across Net framework versions.

If you do not want to test such methods, you can mark them with `[ExcludeCoverage]`, this will disable mutation for this method.

**Reply by psfinaki** [reply]
> I do have unit tests for the couple for GetHashCode overrides, if that is your question.

Can you point to some example please? :)

> If you do not want to test such methods, you can mark them with [ExcludeCoverage], this will disable mutation for this method.

Yeah that's probably a way to go then.

**Reply by dupdob** [reply]
Direct check (this is a massive test for a simple class)
https://github.com/dupdob/NFluent/blob/master/code/tests/NFluent.Tests/DurationTests.cs#L111

Indirect check
https://github.com/dupdob/NFluent/blob/f34e412a385243b6da07e563b33585413af62555/code/tests/NFluent.Tests/EqualRelatedTests.cs#L418

**Reply by psfinaki** [reply]
Thanks for the reference. Yeah the fact that values differ for different frameworks is a bit annoying.
